cmd/juju/cloud/add.go

```diff
@@ -20,6 +20,14 @@ import (
 	"github.com/juju/juju/environs"
 )
 
+type CloudQueryer interface {
+	ParseCloudMetadataFile(path string) (map[string]cloud.Cloud, error)
```

Comment:
If we had the fake call functions from within the `cloud/` package, the boundaries of our tests are then different from that of our code; i.e. we'd be testing something that may not actually be happening. Not only is checking that it's calling `Parse` good enough, it's the only logically correct thing to do here. Likewise, I would not expect the `cloud/` tests to test an interactive add-cloud run; only that when passed data `x` into it's functions, it returns data `y`. The correctness of the add-cloud command when utilized with the `cloud/` package is both an emergent property of both test suites, and an explicit property of a functional test.

Functional tests are about testing that all layers are put together properly and actually work. This is the appropriate place to test that an interactive add-cloud works as intended.

Aside from the logical correctness, there are practical implications as to why you don't want to do functional/full-stack tests here. They are arguably no more difficult to write, but they are significantly more difficult to maintain. Again, please see that link about the testing pyramid. There are fewer functional/CI tests for a reason. Our current tests offer further evidence of this.